apiserver/facades/controller/caasapplicationprovisioner/provisioner.go

```diff
@@ -678,3 +729,520 @@ func readDockerImageResource(reader io.Reader) (params.DockerImageInfo, error) {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		app, err := a.state.Application(appTag.Id())
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		if app.Life() != state.Alive {
+			// We ignore any updates for dying applications.
+			logger.Debugf("ignoring unit updates for dying application: %v", app.Name())
+			continue
+		}
+		appStatus := appUpdate.Status
+		if appStatus.Status != "" && appStatus.Status != status.Unknown {
+			now := a.clock.Now()
+			err = app.SetStatus(status.StatusInfo{
+				Status:  appStatus.Status,
+				Message: appStatus.Info,
+				Data:    appStatus.Data,
+				Since:   &now,
+			})
+			if err != nil {
+				result.Results[i].Error = apiservererrors.ServerError(err)
+				continue
+			}
+		}
+		appUnitInfo, err := a.updateUnitsFromCloud(app, appUpdate.Units)
+		if err != nil {
+			// Mask any not found errors as the worker (caller) treats them specially
+			// and they are not relevant here.
+			result.Results[i].Error = apiservererrors.ServerError(errors.Mask(err))
+		}
+
+		// Errors from SetScale will also include unit info.
+		if appUnitInfo != nil {
+			result.Results[i].Info = &params.UpdateApplicationUnitsInfo{
+				Units: appUnitInfo,
+			}
+		}
+	}
+	return result, nil
+}
+
+type filesystemInfo struct {
+	unitTag      names.UnitTag
+	providerId   string
+	mountPoint   string
+	readOnly     bool
+	size         uint64
+	filesystemId string
+}
+
+type volumeInfo struct {
+	unitTag    names.UnitTag
+	providerId string
+	readOnly   bool
+	persistent bool
+	size       uint64
+	volumeId   string
+}
+
+func (a *API) updateUnitsFromCloud(app Application, unitUpdates []params.ApplicationUnitParams) ([]params.ApplicationUnitInfo, error) {
+	logger.Debugf("unit updates: %#v", unitUpdates)
+
+	m, err := a.state.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var providerIds []string
+	for _, u := range unitUpdates {
+		providerIds = append(providerIds, u.ProviderId)
+	}
+	containers, err := m.Containers(providerIds...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	units, err := app.AllUnits()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	unitByTag := make(map[string]Unit)
+	for _, v := range units {
+		unitByTag[v.Tag().String()] = v
+	}
+	unitByProviderID := make(map[string]Unit)
+	for _, v := range containers {
+		tag := names.NewUnitTag(v.Unit())
+		unit, ok := unitByTag[tag.String()]
+		if !ok {
+			return nil, errors.NotFoundf("unit %q with provider id %q", tag, v.ProviderId())
+		}
+		unitByProviderID[v.ProviderId()] = unit
+	}
+
+	filesystemUpdates := make(map[string]filesystemInfo)
+	filesystemStatus := make(map[string]status.StatusInfo)
+	volumeUpdates := make(map[string]volumeInfo)
+	volumeStatus := make(map[string]status.StatusInfo)
+
+	processFilesystemParams := func(processedFilesystemIds set.Strings, unitTag names.UnitTag, unitParams params.ApplicationUnitParams) error {
```

Comment:
There's a *lot* of code here copied across from caasunitprovisioner facade.
I wonder if we want to have a "common" package and share the code between the 2 implementations. Or do we expect things to diverge significantly.

Reply:
I think they will diverge, just not sure by how much.